sktime/forecasting/base/_base.py

```diff
@@ -591,30 +592,156 @@ def _set_y_X(self, y, X=None, enforce_index_type=None):
+        TypeError if self.get_tag("X-y-must-have-same-index") is True
+            and the index set of X is not a super-set of the index set of y
+
+        Writes to self
+        --------------
+        _y_mtype_last_seen : str, mtype of y
+        _converter_store_y : dict, metadata from conversion for back-conversion
+        """
+        # input checks and minor coercions on X, y
+        ###########################################
+
+        enforce_univariate = self.get_tag("scitype:y") == "univariate"
+        enforce_multivariate = self.get_tag("scitype:y") == "multivariate"
+        enforce_index_type = self.get_tag("enforce_index_type")
+
+        # checking y
+        if y is not None:
+            check_y_args = {
+                "enforce_univariate": enforce_univariate,
+                "enforce_multivariate": enforce_multivariate,
+                "enforce_index_type": enforce_index_type,
+                "allow_None": False,
+                "allow_empty": True,
+            }
+
+            y = check_series(y, **check_y_args, var_name="y")
+
+            self._y_mtype_last_seen = mtype(y)
+        # end checking y
+
+        # checking X
+        if X is not None:
+            X = check_series(X, enforce_index_type=enforce_index_type, var_name="X")
+            if self.get_tag("X-y-must-have-same-index"):
+                check_equal_time_index(X, y)
+        # end checking X
+
+        # convert X & y to supported inner type, if necessary
+        #####################################################
+
+        # retrieve supported mtypes
+
+        # convert X and y to a supported internal mtype
+        #  it X/y mtype is already supported, no conversion takes place
+        #  if X/y is None, then no conversion takes place (returns None)
+        y_inner_mtype = self.get_tag("y_inner_mtype")
+        y_inner = convert_to(
+            y,
+            to_type=y_inner_mtype,
+            as_scitype="Series",  # we are dealing with series
+            store=self._converter_store_y,
+        )
+
+        X_inner_mtype = self.get_tag("X_inner_mtype")
+        X_inner = convert_to(
+            X,
+            to_type=X_inner_mtype,
+            as_scitype="Series",  # we are dealing with series
+        )
+
+        return X_inner, y_inner
+
+    def _check_X(self, X=None):
+        """Shorthand for _check_X_y with one argument X, see _check_X_y."""
+        return self._check_X_y(X=X)[0]
+
     def _update_X(self, X, enforce_index_type=None):
         if X is not None:
             X = check_X(X, enforce_index_type=enforce_index_type)
             if X is len(X) > 0:
                 self._X = X.combine_first(self._X)
 
     def _update_y_X(self, y, X=None, enforce_index_type=None):
-        """Update training data.
+        """Update internal memory of seen training data.
 
         Parameters
         ----------
-        y : pd.Series or pd.DataFrame
+        y : pd.Series, pd.DataFrame, or nd.nparray (1D or 2D)
```

Comment:
Does that mean we now accept numpy arrays internally?

Reply:
yes, since 0.7.something, same version that added `pd.DataFrame` support.
The docstrings have just not been updated - in fact, many of them still referred to `pd.Series` only.

Reply:
I would recommend to update our tests too, to test for inputs of all the types and different number of columns.
I would, however, strongly recommend to adopt the improved fixtures logic in #1407 - this will make it much easier to create X/y input fixtures dependent on estimator tags.

Reply:
(I be strongly against adding more nested loops to the current test logic, that architecture seems quite brittle to me)

Reply:
Oh I must have missed that, I thought we had agreed to internally only support `pd.DataFrame` or `pd.Series`. Could you point me to the PR where this was added?

Reply:
If you'd like my view on the history here:

an early discussion on an "is" state where we are simultaneously using `np.array` and the two `pandas` types for time series is in STEP05, here:
https://github.com/sktime/enhancement-proposals/blob/main/steps/05_scitype_based_IO_checks/step.md

We never had proper "support" for `np.array` since we did not have proper API documentation (including assumes/promises on input/output types) - but as the STEP explains, there has been inconsistent use or "support" throughout the codebase at least since late 2020, probably since 2019.

I think we can't localize this to a PR, but it's more an "organically grown" thing where different contributors would make different assumptions about the target types to support.

The first PR which properly documents the different concrete data containers in use is my PR #1232, and I think it is really the first proper attempt at consolidating the situation. This was done based on sifting through the codebase "as-is", and uncovered that we were using 3 mtypes for individual time series, and 7 mtypes for time series panels.

The observation - made before me documenting the "is" state - is what motivated the creation of the `datatypes` module.

Reply:
at the "end" of this history:

in this PR, besides refactoring the checking logic, I'm trying to move the docstrings to professional grade quality, including clear statements on assumes/guarantees.

As a reviewer, @mloning, I think you shouldn't fault this PR for any property of the "is" state that is not changed by the PR, e.g., the possibility to pass `numpy` objects. You can, of course, ask for making less promises in the docstrings, or discuss formatting issues.